src/main/scala/ir/Program.scala

```diff
@@ -111,25 +170,105 @@ class Program(var procedures: ArrayBuffer[Procedure], var mainProcedure: Procedu
 
 }
 
-class Procedure(
-    var name: String,
-    var address: Option[Int],
-    var blocks: ArrayBuffer[Block],
-    var in: ArrayBuffer[Parameter],
-    var out: ArrayBuffer[Parameter]
-) {
-  def calls: Set[Procedure] = blocks.flatMap(_.calls).toSet
+class Procedure private (
+                  var name: String,
+                  var address: Option[Int],
+                  var entryBlock: Option[Block],
+                  private val _blocks: mutable.HashSet[Block],
+                  var in: ArrayBuffer[Parameter],
+                  var out: ArrayBuffer[Parameter],
+                ) {
+  /* First block executed when the procedure begins */
+  /* returnBlock: Single return point for all returns from the procedure to its caller; always defined but only included
+   * in `blocks` if the procedure contains any real blocks.*/
+  val returnBlock: Block = new Block(name + "_return", None, List(), new IndirectCall(Register("R30", BitVecType(64)), None, Some(name)))
+  returnBlock.setParent(this)
+
+  private var _callers = new mutable.HashSet[Call]
+
+  def this(name: String, address: Option[Int] = None , entryBlock: Option[Block] = None, blocks: IterableOnce[Block] = ArrayBuffer(), in: IterableOnce[Parameter] = ArrayBuffer(), out: IterableOnce[Parameter] = ArrayBuffer()) = {
+    this(name, address, entryBlock, mutable.HashSet.from(blocks), ArrayBuffer.from(in), ArrayBuffer.from(out))
+  }
+
   override def toString: String = {
     s"Procedure $name at ${address.getOrElse("None")} with ${blocks.size} blocks and ${in.size} in and ${out.size} out parameters"
   }
+
+  def calls: Set[Procedure] = blocks.iterator.flatMap(_.calls).toSet
+
+  /**
+   * Horrible, compensating for not storing the blocks in-order and storing initialBlock and returnBlock separately.
+   * @return
+   */
+  def blocks: Seq[Block] =
+    (entryBlock match
+      case Some(b) if _blocks.nonEmpty => Seq(b) ++ _blocks.filter(x => x ne b).toSeq
+      case _ => _blocks.toSeq)
+     ++ (if _blocks.nonEmpty then Seq(returnBlock) else Seq())
```

Comment:
Why does returnBlock need to be stored separately? This also would seem to be non-deterministic, since `_blocks` is a Set, which is very undesirable.

Having this be reconstructed every time `Procedure.blocks` is called seems terribly inefficient. It should instead just provide an iterator over `_blocks`?

Reply:
Its a set for O(1) insertion removal if we modify the IL. My understanding is the block order does not have any meaning beyond the first block, Set's order is not defined but is deterministic it seems. I don't think its neccessary for return block to be stored separately, a few things can probably be cleaned up by just inserting it when the transformation is done.

Reply:
The block order doesn't have any meaning beyond the first block, but using a Set does mean the output order is fragile which is very undesirable for being able to track changes to the output. It also makes the output harder to follow if we further lose correspondence to the original BAP input.

We don't even really care about arbitrary removal right now - the only time blocks are removed at present is when external methods are stubbed out (removing all blocks at once), and anything else is likely to involve iterating over all blocks, so we could really just use a ListBuffer to get O(1) insertion/removal.